packages/react-components/react-motion-preview/src/hooks/useMotion.ts

```diff
@@ -0,0 +1,263 @@
+
+    if (skipAnimation) {
+      return onUnmount;
+    }
+
+    processAnimation(() => {
+      setType(presence ? 'entered' : 'exited');
+      setDelayedAnimationFrame(() => setType(presence ? 'idle' : 'unmounted'));
+    });
+
+    return onUnmount;
+  }, [
+    cancelActiveAnimationFrame,
+    cancelDelayedAnimationFrame,
+    presence,
+    processAnimation,
+    setActiveAnimationFrame,
+    setDelayedAnimationFrame,
+  ]);
+
+  React.useEffect(() => {
+    skipAnimationOnFirstRender.current = false;
+  }, []);
+
+  return React.useMemo(() => {
+    const canRender = () => type !== 'unmounted';
+    const isActive = () => active;
+
+    return {
+      ref,
+      type,
+      canRender,
+      isActive,
+      hasInternalMotion: true,
+    };
+  }, [active, ref, type]);
+}
+
+/**
+ * Returns a default motion state.
+ */
+export function getDefaultMotionState<Element extends HTMLElement>(): MotionState<Element> {
+  return {
+    ref: React.createRef<Element>(),
+    type: 'unmounted',
+    isActive: () => false,
+    canRender: () => false,
+  };
+}
+
+/**
+ * Hook to manage the presence of an element in the DOM based on its CSS transition/animation state.
+ *
+ * @param props - Motion props to manage the presence of an element in the DOM
+ * @param options - Motion options to configure the hook
+ */
+export function useMotion<Element extends HTMLElement>(
+  shorthand: MotionShorthand<Element>,
+  options?: UseMotionOptions,
+): MotionState<Element> {
+  /**
+   * Heads up!
+   * This hook returns a Motion but also accepts Motion as an argument.
+   * In case the hook is called with a Motion as argument, we don't need to perform the expensive computation of the
+   * motion state and can just return the motion value as is. This is intentional as it allows others to use the hook
+   * on their side without having to worry about the performance impact of the hook.
+   */
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  return useIsMotion(shorthand) ? shorthand : useMotionPresence(shorthand, options);
+}
+
+/**
+ * @internal
+ *
+ * This method emits a warning if the hook is called with
+ * a different typeof of shorthand on subsequent renders,
+ * since this can lead breaking the rules of hooks.
+ *
+ * It also return a boolean indicating whether the shorthand is a motion object.
+ */
+export function useIsMotion<Element extends HTMLElement>(
+  shorthand: MotionShorthand<Element>,
+): shorthand is MotionState<Element> {
+  const previousShorthand = usePrevious(shorthand);
+
+  /**
+   * Heads up!
+   * We don't want these warnings in production even though it is against native behavior
+   */
+  if (process.env.NODE_ENV !== 'production') {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    React.useEffect(() => {
+      if (previousShorthand !== null && typeof previousShorthand !== typeof shorthand) {
+        const stringifyShorthand = (value: MotionShorthand<Element>) => JSON.stringify(value, null, 2);
```

Comment:
nit: you can move this to global scope, no need to create new fn on every invocation.

Reply:
Actually, now that I think about it, you only use this when it's not production environment, so it doesn't really matter 😅, sorry for the uneccessary nit